packages/react-api/src/hooks/useGeojsonFeatures.js

```diff
@@ -22,6 +23,8 @@ export default function useGeojsonFeatures({
 
   const sourceId = source?.id;
 
+  useForeignFilter(source);
```

Comment:
The Hook approach has some problems.
- You won't be able to move this logic to builder because the feature calculation happend in Sagas. Take a look [here](https://github.com/CartoDB/cloud-native/blob/main/workspace-www/src/features/builder/state/WidgetsSagas.ts#L357). The only possibility would be to copy and adapt this logic in Builder and is would be hard to maintain .

- I don't like the idea of having a `useEffect` to add and remove the `IN` filter. The workflow is going to be difficult to follow because the `executeTask(sourceId, Methods.GEOJSON_FEATURES, {` is executed at some point and the `useEffect` is executed at another moment.

- I know that the initial approach was to add `IN` filter but thinking about the builder use case, imagine that you have widgets linked in viewport mode and you change any widget to global mode. As you have an `IN` filter created from the link this filter will be sent to the server.

- I would like to propose another approach. I think that we could move the logic to the widget model level. Check this [line](https://github.com/CartoDB/carto-react/blob/master/packages/react-widgets/src/models/CategoryModel.js#L13). Just before to execute the `executeTask` I would execute the `Methods.FILTERED_FEATURES_VALUES` and with the returned values I would refresh the `source.filters` with destructuring (to no modify the original object in the store). I know that is less efficient because we execute  `FILTERED_FEATURES_VALUES` for each widget but the code would be cleaner, automatically would work in Builder, and probably will solve a lot of use cases. What do you think?

Reply:
Thanks for the review.
I'll try with the approach you propose, probably in another branch.

I understand the problem with builder… I forgot that builder doesn't use that much of C4R actually, which is a pity.

> I know that the initial approach was to add IN filter but thinking about the builder use case, imagine that you have widgets linked in viewport mode and you change any widget to global mode. As you have an IN filter created from the link this filter will be sent to the server.

In this case the other hook dedicated to global mode was replacing the filter with the expression for SQL filtering.

> I don't like the idea of having a useEffect to add and remove the IN filter. The workflow is going to be difficult to follow because the executeTask(sourceId, Methods.GEOJSON_FEATURES, { is executed at some point and the useEffect is executed at another moment.

My point was to work at the dataset/feature level, using the existing filter capabilities because I thought (and I still think in some way) they are the right tool. I don't think it would be more difficult to follow than adding special cases at the widget level. 

But once again I'm probably overlooking a lot of implications. And the builder stuff is a real block.

Reply:
Thinking more about the "add IN filter" approach I think that is going to be difficult to maintain because we need to take into account this special filter in some places. Some examples: 
- The `clear filter` action should not remove this IN filter
- In global mode, we also need to ignore this filter to not include the list of `IDs` in the request
- If a user creates a widget using the same column that we use to link we could have problems. Currently, we can differentiate it by the `owner` property but it is a very weak solution.
- If the user removes the linked source we need to remove the IN filter in other sources.
- If in the future we develop more functionalities related to filters like the first two (clear filter and global modal) we're going to have more "exceptions" like in these first two use cases.

Reply:
@padawannn so… I keep coming back with new doubts about the tech approach.

The main thing about not working with real filters is that we don't benefit from the existing system built around them, especially to filter features on the map through Deck.gl datafilter extension.
That means we have to add stuff in this code path too, making maintenance more complicated.

For example, I've tried your proposition here:
> Check this [line](https://github.com/CartoDB/carto-react/blob/master/packages/react-widgets/src/models/CategoryModel.js#L13). Just before to execute the executeTask I would execute the Methods.FILTERED_FEATURES_VALUES and with the returned values I would refresh the source.filters with destructuring (to no modify the original object in the store).

It does the job for the widget result, but it has two main drawbacks:
 - it needs to be duplicated in every widget model
 - it doesn't filter features on the map, which should always reflect the features taken into account in the widget
That's why I think we should work as low-level as possible (at datasource/worker level) and not at the widget level.

Otherwise, I've tried another non-filter approach which is to add a new worker action specific to foreign source, to set foreign values/filtering column. Then the worker can filter when needed and it doesn't conflict with user-set filters, which is useful for clear action, etc. But it has the same drawback as above, filtering of features for the map layer has to be done separately. 

I always come back to a "real" filter that can be applied like the others, benefiting from the whole existing filter workflow.

Now regarding the points you raised above, I think most of them are related to how this auto filter would conflict with user-set ones. I think this could be solved by declaring a new type of filter, let's say `FOREIGN_IN`. That way it could 'cohabit' with another `IN` filter set on the same column, be treated like `IN` in terms of actual implementation, but be easily distinguished when needed (ex: when clearing filters, or for global mode)
I don't think it would be more complicated that all the exceptions that currently exist with the time filters…

What do you think?
I'm 100% sure I'm overlooking obvious blocks here… but honestly otherwise I feel stuck.

Reply:
I forgot the map filtering in my previous comment so you're right, the model approach is not valid. Regarding your proposal about declaring a new type of filter, I think that is a very good idea. Only a suggestion:
In c4r the only way to refresh these filter values is using a useEffect as you did [here](https://github.com/CartoDB/carto-react/pull/468/files#diff-d2205706054f73460f419c20cfaee2e46875c9555d7fe39084486cf27d7ceb58R28) and as we talked, we cannot use this useEffect in builder so we will have to move this logic to `sagas`. Is not a problem because we've already done it but my recommendation is to try to create utils functions and redux selectors that we can use in builder sagas. For example `packages/react-api/src/hooks/dataFilterExtensionUtil.js` was create to be able to use easily in builder